### PR TITLE
feat(react-router): add a `useLocation` hook to get the current `ParsedLocation`

### PR DESCRIPTION
**Pull request by MrPancakes39**
The `useLocation` is a useful hook in [_React Router_](https://reactrouter.com/en/main/hooks/use-location).
So I decided to add it to the _TanStack Router_ too.

### CODE REVIEW

File: packages/react-router/src/useLocation.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { useRouter } from "./useRouter";
+import type { ParsedLocation } from "./location";
+import type { FullSearchSchema } from "./routeInfo";
+import type { AnyRoute } from "./route";
+
+
+export function useLocation<TRouteTree extends AnyRoute = AnyRoute>(): ParsedLocation<FullSearchSchema<TRouteTree>> {
+  const router = useRouter();
```
Comment (SeanCassiere) [inline packages/react-router/src/useLocation.ts:8]: For perf, it'd be better if the selector in the useRouterState hook was used.

```tsx
const location = useRouterState({ select: (s) => s.location })
```

File: docs/framework/react/api/router/useLocationHook.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+---
+id: useLocationHook
+title: useLocation hook
+---
+
+The `useLocation` method is a hook that returns the current [`location`](./ParsedLocationType.md) object. This hook is useful for when you want to perform some side effect whenever the current location changes.
+
+## useLocation returns
+
+- The current [`location`](./ParsedLocationType.md) object.
```
Comment (SeanCassiere) [inline docs/framework/react/api/router/useLocationHook.md:10]: The `.md` extension shouldn't be used.

File: docs/framework/react/api/router/useLocationHook.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+---
+id: useLocationHook
+title: useLocation hook
+---
+
+The `useLocation` method is a hook that returns the current [`location`](./ParsedLocationType.md) object. This hook is useful for when you want to perform some side effect whenever the current location changes.
```
Comment (SeanCassiere) [inline docs/framework/react/api/router/useLocationHook.md:6]: The `.md` extension shouldn't be used.

**Review by SeanCassiere — CHANGES_REQUESTED** [review]
Some tests would also be needed here. You can see the setup in `packages/react-router/tests/router.test.tsx` and grab what you need from there.

**Comment by SeanCassiere** [comment]
Not sure if we are going to add this hook into the library or not, but this PR would need some changes should we decide to add it in.

File: packages/react-router/src/useLocation.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { useRouter } from "./useRouter";
+import type { ParsedLocation } from "./location";
+import type { FullSearchSchema } from "./routeInfo";
+import type { AnyRoute } from "./route";
+
+
+export function useLocation<TRouteTree extends AnyRoute = AnyRoute>(): ParsedLocation<FullSearchSchema<TRouteTree>> {
```
Comment (SeanCassiere) [inline packages/react-router/src/useLocation.ts:7]: Should we be supporting the select syntax here? Like

```tsx
const pathname = useLocation({ select: s => s.pathname })
```

Comment (tannerlinsley) [reply packages/react-router/src/useLocation.ts:7]: Yes probably

**Review by SeanCassiere — APPROVED** [review]

**Comment by SeanCassiere** [comment]
Good to go.

**Comment by MrPancakes39** [comment]
I apologize for not seeing the needed changes sooner and resolving them myself.
Thank you @SeanCassiere for your consideration and work.

**Pull request merged**